src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -0,0 +1,597 @@
+        /// <exception cref="MaxReorgViolationException">Thrown in case maximum reorganization rule is violated.</exception>
+        /// <exception cref="ConnectHeaderException">Thrown if it wasn't possible to connect the first new header.</exception>
+        private List<ChainedHeader> CreateNewHeaders(List<BlockHeader> headers)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(headers), nameof(headers.Count), headers.Count);
+
+            if (!this.FindFirstNewHeader(headers, out int newHeaderPosition))
+            {
+                this.logger.LogTrace("(-)[NO_NEW_HEADERS_FOUND]:null");
+                return null;
+            }
+
+            ChainedHeader previousChainedHeader = this.chainedHeadersByHash.TryGet(headers[newHeaderPosition].HashPrevBlock);
+            if (previousChainedHeader == null)
+            {
+                this.logger.LogTrace("Previous hash `{0}` of block hash `{1}` was not found.", headers[newHeaderPosition].GetHash(), headers[newHeaderPosition].HashPrevBlock);
+                this.logger.LogTrace("(-)[PREVIOUS_HEADER_NOT_FOUND]");
+                throw new ConnectHeaderException();
+            }
+
+            var newChainedHeaders = new List<ChainedHeader>();
+
+            ChainedHeader newChainedHeader = this.CreateAndValidateNewChainedHeader(newChainedHeaders, headers[newHeaderPosition], previousChainedHeader);
+            newHeaderPosition++;
+            this.logger.LogTrace("New chained header was added to the tree '{0}'.", newChainedHeader);
+
+            try
+            {
+                this.CheckMaxReorgRuleViolated(newChainedHeader);
+
+                previousChainedHeader = newChainedHeader;
+
+                for (; newHeaderPosition < headers.Count; newHeaderPosition++)
+                {
+                    newChainedHeader = this.CreateAndValidateNewChainedHeader(newChainedHeaders, headers[newHeaderPosition], previousChainedHeader);
+                    this.logger.LogTrace("New chained header was added to the tree '{0}'.", newChainedHeader);
+
+                    previousChainedHeader = newChainedHeader;
+                }
+            }
+            catch
+            {
+                // Undo changes to the tree. This is necessary because the peer claim wasn't set to the last header yet.
+                // So in case of peer disconnection this branch wouldn't be removed.
+                this.RemoveUnclaimedBranch(newChainedHeader);
+
+                this.logger.LogTrace("(-)[VALIDATION_FAILED]");
+                throw;
+            }
+
+            this.logger.LogTrace("(-):*.{0}:{1}", nameof(newChainedHeaders.Count), newChainedHeaders.Count);
+            return newChainedHeaders;
+        }
+
+        private ChainedHeader CreateAndValidateNewChainedHeader(List<ChainedHeader> newChainedHeaders, BlockHeader currentBlockHeader, ChainedHeader previousChainedHeader)
```

Comment:
This method should not accept list of chain headers as its argument. The change to the latest should be made by the caller of this method.

Reply:
Why not? its sort of a method that will take away noise form the caller, and will only be called in one place.

Reply:
It makes the code unreadable. That logic belongs to the caller. The caller owns the list. the caller works with the list. The caller should populate the list.

This method does more than it should. I very strongly believe the argument should be removed.